two fixes needed for installation on MacOS 10.13.6 with python3.7

I needed Qt5 to build Slicer and SlicerAstroApp and used this distribution as other installations failed. When building and installing Qt5 I ran into two errors I needed to fix. The messages and fixes I made are given below.

_**error 1:**_

```
ninja  -C /Users/vdhulst/Software/qt-everywhere-src-5.15.0/qtwebengine/src/core/release QtWebEngineCore
ninja: Entering directory `/Users/vdhulst/Software/qt-everywhere-src-5.15.0/qtwebengine/src/core/release'
[1/15903] LIBTOOL-STATIC obj/third_party/SPIRV-Tools/src/libspvtools.a
FAILED: obj/third_party/SPIRV-Tools/src/libspvtools.a 
rm -f obj/third_party/SPIRV-Tools/src/libspvtools.a && TOOL_VERSION=1588774889 python ../../3rdparty/chromium/build/toolchain/mac/filter_libtool.py /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool -static  -o obj/third_party/SPIRV-Tools/src/libspvtools.a -filelist obj/third_party/SPIRV-Tools/src/libspvtools.a.rsp
Traceback (most recent call last):
  File "../../3rdparty/chromium/build/toolchain/mac/filter_libtool.py", line 56, in <module>
    sys.exit(Main(sys.argv[1:]))
  File "../../3rdparty/chromium/build/toolchain/mac/filter_libtool.py", line 43, in Main
    if not IsBlacklistedLine(line):
  File "../../3rdparty/chromium/build/toolchain/mac/filter_libtool.py", line 28, in IsBlacklistedLine
    if pattern.match(line):
TypeError: cannot use a string pattern on a bytes-like object
[10/15903] CXX obj/third_party/SPIRV-Tools/src/spvtools_opt/amd_ext_to_khr.o
ninja: build stopped: subcommand failed.
make[4]: *** [run_ninja] Error 1
make[3]: *** [sub-gn_run-pro-make_first] Error 2
make[2]: *** [sub-core-make_first] Error 2
make[1]: *** [sub-src-make_first] Error 2
make: *** [module-qtwebengine-make_first] Error 2
```

fixed by changing: `if pattern.match(line):` into `if pattern.match(str(line)):`
in filter_libtool.py

**_error 2:_**

```
In file included from gen/third_party/blink/renderer/core/core_generated_jumbo_11.cc:8:
/Users/vdhulst/Software/qt-everywhere-src-5.15.0/qtwebengine/src/core/release/gen/third_party/blink/renderer/core/xpath_grammar.cc:124:10: fatal error: 'xpath_grammar.hh' file not found
#include "xpath_grammar.hh"
```

 fixed by changing `#include "xpath_grammar.hh"` into `#include "xpath_grammar.h"`
 in xpath_grammar.cc

--- a/Build-qt.sh
+++ b/Build-qt.sh
@@ -370,10 +370,6 @@ then
 fi
 cd $src_dir
 
-# Options used to mimic the homebrew packaging of Qt5
-#qt_homebrew_package_options="-system-zlib -qt-libpng -qt-libjpeg -qt-freetype -qt-pcre -dbus-runtime -proprietary-codecs"
-qt_build_mode="-silent"
-qt_build_mode="-verbose"
 
 # NOTE:  C++14 is needed to support QtWebEngine from chromium
 ./configure $qt_install_dir_options                           \

--- a/Build-qt.sh
+++ b/Build-qt.sh
@@ -2,6 +2,9 @@
 set -ex
 set -o pipefail
 
+script_dir=$(cd $(dirname $0) || exit 1; pwd)
+
+
 #
 # Configuration
 #
@@ -358,6 +361,12 @@ qt_install_dir_options="-prefix $install_dir"
 if [[ ! -d $src_dir ]]
 then
   tar --no-same-owner -xf $deps_dir/$qt_archive
+
+  # Apply patches
+  pushd $src_dir
+  git apply --ignore-whitespace $script_dir/patches/*.patch
+  popd
+
 fi
 cd $src_dir
 

--- a/Build-qt.sh
+++ b/Build-qt.sh
@@ -348,6 +348,28 @@ then
 fi
 cd ..
 
+# Build Python 2.7: required to build QtWebEngine and QtPdf in Qt 5.15.x
+echo "Build Python 2.7"
+
+cwd=$(pwd)
+
+mkdir -p python-cmake-buildsystem-build
+if [[ ! -d python-cmake-buildsystem ]]
+then
+  git clone https://github.com/python-cmake-buildsystem/python-cmake-buildsystem.git
+fi
+cd python-cmake-buildsystem-build
+$cmake \
+  -DCMAKE_BUILD_TYPE:STRING=Release \
+  -DCMAKE_INSTALL_PREFIX:PATH=$cwd/python-cmake-buildsystem-install \
+  -DPYTHON_VERSION:STRING=2.7.15 \
+  -DBUILD_LIBPYTHON_SHARED:BOOL=ON \
+  -DENABLE_SSL:BOOL=OFF \
+  ../python-cmake-buildsystem
+
+$cmake --build $cwd/python-cmake-buildsystem-build --config Release --target install -- -j$nbthreads
+cd ..
+
 popd
 
 # Build Qt
@@ -370,6 +392,7 @@ then
 fi
 cd $src_dir
 
+export PATH=$deps_dir/python-cmake-buildsystem-install/bin:$PATH
 
 # NOTE:  C++14 is needed to support QtWebEngine from chromium
 ./configure $qt_install_dir_options                           \

--- a/patches/0001-8467beddb-add-missing-macos-header-file-that-was-indirectly-included.patch
+++ b/patches/0001-8467beddb-add-missing-macos-header-file-that-was-indirectly-included.patch
@@ -1,0 +1,26 @@
+From 8467beddb7239cc213ae13900fa30e3d26df5e78 Mon Sep 17 00:00:00 2001
+From: =?UTF-8?q?=C3=98ystein=20Heskestad?= <oystein.heskestad@qt.io>
+Date: Wed, 27 Oct 2021 13:07:46 +0200
+Subject: [PATCH] Add missing macOS header file that was indirectly included
+ before
+
+Change-Id: I4d4c7d4f957fc36dea5e06eb6d661aeecf6385f1
+(cherry picked from commit dece6f5840463ae2ddf927d65eb1b3680e34a547)
+Reviewed-by: Volker Hilsheimer <volker.hilsheimer@qt.io>
+---
+ src/plugins/platforms/cocoa/qiosurfacegraphicsbuffer.h | 2 ++
+ 1 file changed, 2 insertions(+)
+
+diff --git a/qtbase/src/plugins/platforms/cocoa/qiosurfacegraphicsbuffer.h b/qtbase/src/plugins/platforms/cocoa/qiosurfacegraphicsbuffer.h
+index dd3ebca4a65..ee7ac5c0143 100644
+--- a/qtbase/src/plugins/platforms/cocoa/qiosurfacegraphicsbuffer.h
++++ b/qtbase/src/plugins/platforms/cocoa/qiosurfacegraphicsbuffer.h
+@@ -43,6 +43,8 @@
+ #include <qpa/qplatformgraphicsbuffer.h>
+ #include <private/qcore_mac_p.h>
+ 
++#include <CoreGraphics/CGColorSpace.h>
++
+ QT_BEGIN_NAMESPACE
+ 
+ class QIOSurfaceGraphicsBuffer : public QPlatformGraphicsBuffer

--- a/patches/README.md
+++ b/patches/README.md
@@ -1,0 +1,11 @@
+# Qt Patches
+
+Fixes and improvements to Qt that should first be contributed upstream. The [Qt Contribution Guidelines](https://wiki.qt.io/Qt_Contribution_Guidelines) is a great resource to guide you through the process.
+
+Each patch is documented by adding an entry below. Whenever possible, references to bugs.python.org https://bugreports.qt.io/, corresponding GitHub PRs and any related discussions on forums or mailing lists are also provided.
+
+Before being applied, patches are sorted alphabetically. This ensures that patch starting with 0001- is applied before the one starting with 0002-.
+
+## Patches
+
+* NA

--- a/patches/README.md
+++ b/patches/README.md
@@ -8,4 +8,4 @@ Before being applied, patches are sorted alphabetically. This ensures that patch
 
 ## Patches
 
-* NA
+* [0001-8467beddb-add-missing-macos-header-file-that-was-indirectly-included.patch][0001-8467beddb-add-missing-macos-header-file-that-was-indirectly-included.patch] adapted from [qt/qtbase@8467beddb](https://github.com/qt/qtbase/commit/8467beddb7239cc213ae13900fa30e3d26df5e78). The `qtbase/` sub-directory has been added to the patch.

